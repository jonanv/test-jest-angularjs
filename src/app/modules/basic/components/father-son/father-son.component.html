<p>father-son works!</p>
