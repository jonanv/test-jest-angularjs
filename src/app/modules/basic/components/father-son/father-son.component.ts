import { Component } from '@angular/core';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styles: ``
})
export class FatherSonComponent {

}
