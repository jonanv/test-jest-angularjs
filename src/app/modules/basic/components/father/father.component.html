<h1>Padre</h1>
<button
  class="btn btn-primary"
  (click)="onSetClient('Giovanni');">
  Cargar cliente
</button>

<br><br>

<code
  class="mt-2">
  {{ client | json }}
</code>

<hr>

<app-father-son
  [client]="client"
  (onDeleteClientEmitter)="processOnDeleteClientEmitter();"
  (onUpdateClientEmitter)="processOnUpdateClientEmitter($event);">
</app-father-son>
