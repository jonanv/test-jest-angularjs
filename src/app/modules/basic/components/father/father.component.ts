import { Component } from '@angular/core';

@Component({
  selector: 'app-father',
  templateUrl: './father.component.html',
  styles: ``
})
export class FatherComponent {

}
