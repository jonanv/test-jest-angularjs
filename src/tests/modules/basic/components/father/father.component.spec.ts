import { ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';

// Components
import { FatherComponent } from '../../../../../app/modules/basic/components/father/father.component';
import { FatherSonComponent } from '../../../../../app/modules/basic/components/father-son/father-son.component';

describe('FatherComponent', () => {
  let component: FatherComponent;
  let fixture: ComponentFixture<FatherComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FatherComponent, FatherSonComponent]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement as HTMLElement;
  });

  test('Debe ser creado el componente', () => {
    expect(component).toBeTruthy();
  });

  test('Debe de hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('Debe de hacer match con el snapshot si hay cliente', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot();
  });

  test('Debe de establer el nombre del cliente', () => {
    component.onSetClient('Juan');
    fixture.detectChanges();

    let divCode = compiled.querySelector('.mt-2');
    expect(divCode?.textContent).toContain("name");
    expect(divCode?.textContent).toContain("Juan");
  });

  test('Debe de borrar el cliente si se emite onDeleteClient (hijo)', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query(By.directive(FatherSonComponent));
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance;
    // console.log(sonComponent.client);
    sonComponent.onDeleteClientEmitter.emit();
    expect(component.client).toBeUndefined();
  });

  test('Debe de actualizar el cliente onUpdateCliente', () => {
    component.client = { id: 1, name: 'Pedro' };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query(By.directive(FatherSonComponent));
    const sonComponent: FatherSonComponent = sonDebugElement.componentInstance;
    // console.log(sonComponent.client);
    sonComponent.onUpdateClientEmitter.emit({ id: 2, name: 'Pedro' });
    expect(component.client).toEqual({ id: 2, name: 'Pedro' });

    let divCode = compiled.querySelector('.mt-2');
    expect(divCode?.textContent).toContain("name");
    expect(divCode?.textContent).toContain("Pedro");
  });

});
