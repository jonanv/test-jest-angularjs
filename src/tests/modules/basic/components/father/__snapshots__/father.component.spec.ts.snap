// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent Debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h1>
    Padre
  </h1>
  <button
    class="btn btn-primary"
    data-test="btn-set-client"
  >
     Cargar cliente

  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
     

  </code>
  <hr />
  <app-father-son>
    <h5>
      Hijo
    </h5>
    <code>
      
    </code>
    <br />
    <br />
    <div
      class="d-flex gap-2"
    >
      
      
    </div>
  </app-father-son>
</div>
`;

exports[`FatherComponent Debe de hacer match con el snapshot si hay cliente 1`] = `
<div
  id="root2"
>
  <h1>
    Padre
  </h1>
  <button
    class="btn btn-primary"
    data-test="btn-set-client"
  >
     Cargar cliente

  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
     {
  "id": 1,
  "name": "Juan"
}

  </code>
  <hr />
  <app-father-son>
    <h5>
      Hijo
    </h5>
    <code>
      {
  "id": 1,
  "name": "Juan"
}
    </code>
    <br />
    <br />
    <div
      class="d-flex gap-2"
    >
      <button
        class="btn btn-primary"
        data-test="btn-update-client"
      >
         Cambiar ID 
      </button>
      
      <button
        class="btn btn-danger"
        data-test="btn-delete-client"
      >
         Borrar cliente 
      </button>
      
    </div>
  </app-father-son>
</div>
`;
